core/src/test/scala/unit/kafka/server/HeartbeatRequestTest.scala

```diff
@@ -0,0 +1,199 @@
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.HEARTBEAT.oldestVersion() to ApiKeys.HEARTBEAT.latestVersion(isUnstableApiEnabled)) {
+      val metadata = ConsumerProtocol.serializeSubscription(
+        new ConsumerPartitionAssignor.Subscription(Collections.singletonList("foo"))
+      ).array
+
+      val (leaderMemberId, leaderEpoch) = joinDynamicConsumerGroupWithOldProtocol(
+        groupId = "grp",
+        metadata = metadata,
+        completeRebalance = false
+      )
+
+      // Heartbeat with unknown group id and unknown member id.
+      heartbeat(
+        groupId = "grp-unknown",
+        memberId = "member-id-unknown",
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+
+      // Heartbeat with unknown group id.
+      heartbeat(
+        groupId = "grp-unknown",
+        memberId = leaderMemberId,
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+
+      // Heartbeat with unknown member id.
+      heartbeat(
+        groupId = "grp",
+        memberId = "member-id-unknown",
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+
+      // Heartbeat with unmatched generation id.
+      heartbeat(
+        groupId = "grp",
+        memberId = leaderMemberId,
+        generationId = -1,
+        expectedError = Errors.ILLEGAL_GENERATION,
+        version = version.toShort
+      )
+
+      // Heartbeat COMPLETING_REBALANCE group.
+      heartbeat(
+        groupId = "grp",
+        memberId = leaderMemberId,
+        generationId = leaderEpoch,
+        version = version.toShort
+      )
+
+      syncGroupWithOldProtocol(
+        groupId = "grp",
+        memberId = leaderMemberId,
+        generationId = leaderEpoch,
+        assignments = List(new SyncGroupRequestData.SyncGroupRequestAssignment()
+          .setMemberId(leaderMemberId)
+          .setAssignment(Array[Byte](1))
+        )
+      )
+
+      // Heartbeat STABLE group.
+      heartbeat(
+        groupId = "grp",
+        memberId = leaderMemberId,
+        generationId = leaderEpoch,
+        version = version.toShort
+      )
+
+      // Join the second member.
+      val joinFollowerResponseData = sendJoinRequest(
+        groupId = "grp",
+        metadata = metadata
+      )
+
+      Future {
+        sendJoinRequest(
+          groupId = "grp",
+          memberId = joinFollowerResponseData.memberId,
+          metadata = metadata
+        )
+      }
+
+      TestUtils.waitUntilTrue(() => {
+        val described = describeGroups(groupIds = List("grp"), version = ApiKeys.DESCRIBE_GROUPS.latestVersion(isUnstableApiEnabled))
+        GenericGroupState.PREPARING_REBALANCE.toString == described.head.groupState
+      }, msg = s"The group is not in PREPARING_REBALANCE state.")
```

Comment:
It doesn't work because somehow the code in the previous future tend to be executed later than the following code, so I have to set a waitUntilTrue or sleep for a while to make sure the request in the future is processed first.